Add data widgets for rendering simple (non-Django-model) data

tonysyu opened the pull request:
Marking this as a WIP since this is currently missing examples, documentation, and tests. I'd like to check that this would be a welcome addition before putting in the effort to finish this out.

Basically, this adds a couple of widgets, `ValueList` and `KeyValueList` which can be used to render dashboard widgets from python lists and dictionaries instead of Django models. For example:

```python
from controlcenter import widgets
from controlcenter.utils import DataItem
from django.conf import settings

class DebuggingEndpointsWidget(widgets.ValueList):
    title = 'Debugging Endpoints'

    def get_data(self):
        return [
            DataItem('Healthcheck', url='/private/healthcheck', help_text='App healthcheck report'),
            DataItem('Datadog Dashboard', url='https://example.com'),
            # ...
        ]

class AppInfoWidget(widgets.KeyValueList):
    title = 'App info'

    def get_data(self):
        return {
            'App Version': settings.CURRENT_APP)VERSION,
            'Debug on': settings.DEBUG,
            'Default timezone': settings.TIME_ZONE,
            # ...
        }

```

byashimov commented:
Hi, 
that's a great idea! Any contribution is be appreciated :)
Please, create a new module `widgets.contrib` with your widgets and let me know if I you need any help.

tonysyu commented:
> Please, create a new module widgets.contrib with your widgets
Do you mean I should rename `widgets.simple` to `widgets.contrib` or are you suggesting a new `contrib` sub-package such that the widgets are in `widgets.contrib.simple`

byashimov commented on controlcenter/templatetags/controlcenter_tags.py:
```diff
@@ -205,3 +205,15 @@ def attrlabel(widget, attrname):
+        'rel="noreferrer"',
+        'rel="noopener"',
+    ])
+    return mark_safe('<a {}>{}</a>'.format(attrs, label))
```
Better make one string template:
```python
return mark_safe(
    '<a href="{href}" target="_blank" '
    'rel="noreferrer" rel="noopener">{label}</a>'
    .format(href=url, label=label or url)
)
```

tonysyu commented:
Putting it in `widgets.contrib.simple` sounds good. BTW, the test failures are due to changes in Django 2.1 (and not related to this PR). It seems that the dashboard used for testing use widgets that don't define `template_name`. Before Django 2.1, calling `get_template_name` from the dashboard template when `Widget.template_name` isn't defined suppressed. It seems that this error gets bubbled up in Django 2.1. (I don't see any mention of this in release notes, though)

tonysyu commented:
@byashimov: I've moved the widgets to `widgets.contrib.simple`, added tests, and added examples to the docs. In addition, I added a sample project, which adds quite a bit to the diff. If you'd prefer to take it out, I can easily just revert the last commit.

tonysyu commented:
I don't really understand the results from @codecov-io: When I run the following locally, the new widgets and template tag have 100% coverage:
```
$ coverage run test_project/manage.py test; coverage report -m
```

byashimov commented:
Hi! It's fixed in develop now. Could you rebase your branch, please?

byashimov commented:
Oh, the example project is really awesome and you have done a great job, but there is already the test_project and docs with verbose examples. And I'm not really sure if I will have enough time to make compatibility fixes in future for it too. So if you don't mind I'll accept the feature only. Thanks.

tonysyu commented:
Rebased and tests are passing! Also removed the example project.

tonysyu commented:
@byashimov: Anything else?

merwok commented on controlcenter/templatetags/controlcenter_tags.py:
```diff
@@ -205,3 +205,12 @@ def attrlabel(widget, attrname):
+
+@register.simple_tag
+def external_link(url, label=None):
+    return mark_safe(
```
I think mark_safe is superseded by format_html since Django 1.10 at least.

byashimov commented on controlcenter/templates/controlcenter/dashboard.html:
```diff
@@ -27,6 +27,9 @@
+                                {% if widget.subtitle %}
+                                    <div class="controlcenter__widget__subtitle">{{ widget.subtitle }}</div>
+                                {% endif %}
                                 {% if widget.cache_timeout %}
```
Is there any way to not affect dashboard template?

byashimov commented on controlcenter/widgets/contrib/simple.py:
```diff
@@ -0,0 +1,45 @@
+
+    def get_data(self):
+        raise NotImplementedError("SimpleWidget subclass must implement `get_data`")
+
```
This should be BaseDataItemWidget or something, it can not be used as is.
